libs/sdk-core/src/lsps0/transport.rs

```diff
@@ -0,0 +1,677 @@
+}
+
+#[tonic::async_trait]
+impl<TNotification> NotificationSender for NotificationHandler<TNotification>
+where
+    TNotification: DeserializeOwned + Send,
+{
+    async fn send(&self, msg: JsonRpcNotification) -> Result<(), Error> {
+        let n = match serde_json::from_value::<TNotification>(msg.params) {
+            Ok(n) => n,
+            Err(e) => return Err(Error::Deserialization(e)),
+        };
+
+        match self.tx.send(n).await {
+            Ok(_) => Ok(()),
+            Err(_) => Err(Error::Local(anyhow!("receiver dropped"))),
+        }
+    }
+}
+
+struct ResponseOrError {
+    response: Option<JsonRpcResponse>,
+    error: Option<JsonRpcError>,
+}
+
+/// Transport sends and receives LSPS0 messages to/from remote nodes. One
+/// user node has one Transport.
+pub struct Transport {
+    node: Arc<dyn NodeAPI>,
+    response_handlers: Mutex<HashMap<String, oneshot::Sender<ResponseOrError>>>,
+    notification_handlers: RwLock<HashMap<String, Box<dyn NotificationSender>>>,
+}
+
+impl Transport {
+    #[allow(dead_code)]
+    pub fn new(node: Arc<dyn NodeAPI>) -> Transport {
+        Transport {
+            node,
+            response_handlers: Mutex::new(HashMap::new()),
+            notification_handlers: RwLock::new(HashMap::new()),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn start(&self) -> Result<()> {
+        debug!("starting lsps0 tranport.");
+        let mut stream = self.node.stream_custom_messages().await?;
+        loop {
+            let msg = match stream.recv().await {
+                Some(val) => val,
+                None => break Err(anyhow!("custom message stream ended.")),
+            };
+            if msg.message_type != LSPS0_MESSAGE_TYPE {
+                continue;
+            }
+
+            let v: JsonRpcServerObject = match serde_json::from_slice(&msg.payload) {
+                Ok(v) => v,
+                Err(e) => {
+                    debug!(
+                        "error deserializing lsps0 payload {:?}: {}",
+                        &msg.payload, e
+                    );
+                    continue;
+                }
+            };
+
+            let jsonrpc_version = match &v {
+                JsonRpcServerObject::Notification(n) => &n.jsonrpc,
+                JsonRpcServerObject::Response(n) => &n.jsonrpc,
+                JsonRpcServerObject::Error(n) => &n.jsonrpc,
+            };
+
+            if jsonrpc_version != JSONRPC_VERSION {
+                debug!("error deserializing lsps0 payload {:?}: Invalid jsonrpc version. Expected {:?}.", &msg.payload, JSONRPC_VERSION);
+                continue;
+            }
+
+            match v {
+                JsonRpcServerObject::Notification(n) => {
+                    let id = get_notification_handler_id(&n.method, msg.peer_id.clone());
+                    if let Some(tx) = (*self.notification_handlers.read().await).get(&id) {
+                        match tx.send(n).await {
```

Comment:
The send is inside the lock, perhaps it is better to release the lock immediately after we finished reading from the notification_handlers.

Reply:
Isn't the lock out of scope already here?

Reply:
Yes you are right , the temporary variable is release.